storage/merkle/proof.go

```diff
@@ -0,0 +1,133 @@
+package merkle
+
+import (
+	"bytes"
+
+	"github.com/onflow/flow-go/ledger/common/bitutils"
+)
+
+// Proof captures all data needed for proving inclusion of a single value inserted under key `Key` into the merkle trie
+type Proof struct {
+	// Key used to insert and look up the value
+	Key []byte
+	// Value stored in the trie for the given key
+	Value []byte
+	// IsAShortNode holds bits of data to determine short nodes versus full nodes while tranversing the
+	// trie downward. if the bit is set to 1, it means that we have reached to a short node, and
+	// if is set to 0 means we have reached a full node.
+	IsAShortNode []byte
+	// SkipBits is read when we reach a short node, and the value represents number of bits that were skipped
+	// by the short node (shortNode.count)
+	SkipBits []uint16
+	// InterimHashes is a slice of hash values, every value is read when we reach a full node (hash value of the siblings)
+	InterimHashes [][]byte
+}
+
+// Verify verifies the proof by constructing the hash values bottom up and cross check
+// the constructed root hash with the given one.
+// if the proof is valid it returns true and false otherwise
+func (p *Proof) Verify(expectedRootHash []byte) (bool, error) {
+
+	// first check the key size to be smaller than max allowed
+	if len(p.Key) > maxKeyLength {
+		return false, NewErrorMalformedProoff("key length is larger than max key lenght allowed (%d > %d)", len(p.Key), maxKeyLength)
+	}
+
+	// number of steps
+	steps := len(p.SkipBits) + len(p.InterimHashes)
+	// number of steps should be smaller than max key length
+	if steps > maxKeyLength {
+		return false, NewErrorMalformedProoff("length of SkipBits pluse length of InterimHashes is larger than max key lenght allowed (%d > %d)", steps, maxKeyLength)
+	}
+	// number of steps should be smaller than number of bits in the IsAShortNode
+	if steps > len(p.IsAShortNode)*8 {
```

Comment:
you check here, that `IsAShortNode` is not too short. But you don't check that it is not excessively too long. I could literally send you a `IsAShortNode` that is 2GB in size, where only the leading bytes are used and the rest is filled with garbage. Your algorithm wouldn't detect anything wrong with this proof. 

* we know _precisely_ what length we expect: 
```suggestion
	if len(p.IsAShortNode) != (steps+7)>>3 {
```

Reply:
I'll add it